src/java.base/share/classes/javax/security/auth/login/LoginContext.java

```diff
@@ -288,6 +293,24 @@ public ClassLoader run() {
                     return loader;
                 }
         });
+
+        synchronized(cacheServiceProviders){
+            if (cacheServiceProviders.get(contextClassLoader) == null){
+                if (debug != null)
+                    debug.println("Build ServiceProviders cache for ClassLoader: " + contextClassLoader.getName());
+                @SuppressWarnings("removal")
+                ServiceLoader<LoginModule> sc = AccessController.doPrivileged(
+                        (PrivilegedAction<ServiceLoader<LoginModule>>)
+                                () -> java.util.ServiceLoader.load(
+                                    LoginModule.class, contextClassLoader));
+                Set<Provider<LoginModule>> lmProviders = sc.stream().collect(Collectors.toSet());
```

Comment:
It's probably necessary to put the `collect` call above inside doPriv as well. The actual class loading might happen later.

Reply:
I'm seeing a JCK failure on this change. Will investigate more. At the same time, is it possible to merge this synchronized block into the one below?